### DISCUSSION TITLE
INSFV not converging

### DISCUSSION THREAD

**Discussion by Spudman105** [idaholab/moose#28909] [Q&A Modules: Navier-Stokes]
### Check these boxes if you have followed the posting rules.

- [X] Q&A Navier Stokes is the most appropriate category for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (avoid screenshots if possible, triple back quotes before/after pasted text, etc)

### Question

Hello, I am currently trying to use the Incompressible Navier Stokes Finite Volume module to simulate a shock wave motion moving from one end to another of a cylinder. I realize that there is a chance that this might not be able to be done. At this point I used a velocity equation to control how the loading would be as an input BC. It will run however the linear residual wont converge nor will the non-linear even with really relaxed relative tolerance. Usually it fails due to diverged_breakdown or line_search breakdown. I have tried using preconditioning to see if there is something wrong with the Jacobian but have not come up with any solutions. Below is my code and some of the output.

'''
mu = 1  #incompressible, dynamic viscosity
rho = 8.027 #(g/cm^3), Density
k = 14.6 #W/m k Thermal conductivity
cp = 450 #j/kg K, specific heat

[GlobalParams]

  rhie_chow_user_object = 'rc'
[]

[Mesh]
  [face1]
      type = AnnularMeshGenerator
      rmin = 0 # makes the plane a disk instead of a ring
      rmax = .5 # sets the radius from the center for the outer ring, r=.5 means diameter is 1 which is beam spot size.
      nt = 100 # number of notes along outer ring 
      nr = 5 # number of nodes between the center and the outer radius. (increase later but small now for speed of tests)
  []
  [extrude]
      type = AdvancedExtruderGenerator
      input = face1
      heights = '1 2'
      num_layers = '24 12 ' # the number of between layers from one height to another
      direction = '0 0 1'
      biases = '1 2'
  []
[]

[Variables]
  [vel_x]
    type = INSFVVelocityVariable
  []
  [vel_y]
    type = INSFVVelocityVariable
  []
  [vel_z]
    type = INSFVVelocityVariable
  []
  [pressure]
    type = INSFVPressureVariable
  []
  [T_fluid]
    type = INSFVEnergyVariable
  []
[]

[ICs]
  [ic_vel_x]
    type = ConstantIC
    variable = vel_x
    value = 0
  []
  [ic_vel_y]
    type = ConstantIC
    variable = vel_y
    value = 0
  []
  [ic_vel_z]
    type = ConstantIC
    variable = vel_z
    value = 0
  []
  [ic_pressure]
    type = ConstantIC
    variable = pressure
    value = 1
  []
  [ic_T_fluid]
    type = ConstantIC
    variable = T_fluid
    value = 300
  []
[]

[UserObjects]
  [rc]
    type = INSFVRhieChowInterpolator
    u = vel_x
    v = vel_y
    w = vel_z
    pressure = pressure
  []
[]

[FVKernels]
  [mass]
    type = INSFVMassAdvection
    variable = pressure
    advected_interp_method = 'average'
    rho = ${rho}
  []
  [mass_forcing]
    type = FVBodyForce
    variable = pressure
  []

  [u_advection]
    type = INSFVMomentumAdvection
    variable = vel_x
    advected_interp_method = 'average'
    rho = ${rho}
    momentum_component = 'x'
  []

  [u_viscosity]
    type = INSFVMomentumDiffusion
    variable = vel_x
    mu = ${mu}
    momentum_component = 'x'
  []

  [u_pressure]
    type = INSFVMomentumPressure
    variable = vel_x
    momentum_component = 'x'
    pressure = pressure
  []

  [v_advection]
    type = INSFVMomentumAdvection
    variable = vel_y
    advected_interp_method = 'average'
    rho = ${rho}
    momentum_component = 'y'
  []

  [v_viscosity]
    type = INSFVMomentumDiffusion
    variable = vel_y
    mu = ${mu}
    momentum_component = 'y'
  []

  [v_pressure]
    type = INSFVMomentumPressure
    variable = vel_y
    momentum_component = 'y'
    pressure = pressure
  []

  [w_advection]
    type = INSFVMomentumAdvection
    variable = vel_z
    advected_interp_method = 'average'
    rho = ${rho}
    momentum_component = 'z'
  []

  [w_viscosity]
    type = INSFVMomentumDiffusion
    variable = vel_z
    mu = ${mu}
    momentum_component = 'z'
  []

  [w_pressure]
    type = INSFVMomentumPressure
    variable = vel_z
    momentum_component = 'z'
    pressure = pressure
  []

  [temp_conduction]
    type = FVDiffusion
    coeff = 'k'
    variable = T_fluid
  []

  [temp_advection]
    type = INSFVEnergyAdvection
    variable = T_fluid
    advected_interp_method = 'average'
  []
[]

[FVBCs]
  [Inlet]
    type = INSFVInletVelocityBC
    boundary = '3'
    functor = 'velocity_z'
    variable = vel_z
  []
  [slip_x]
    type = INSFVNaturalFreeSlipBC
    variable = vel_x
    boundary = 'rmax'
    momentum_component = 'x'
  []
  [slip_y]
    type = INSFVNaturalFreeSlipBC
    variable = vel_y
    boundary = 'rmax'
    momentum_component = 'y'
  []
  [slip_z]
    type = INSFVNaturalFreeSlipBC
    variable = vel_z
    boundary = 'rmax'
    momentum_component = 'z'
  []
  [T_hot]
    type = FVDirichletBC
    variable = T_fluid
    boundary = '1'
    value = 310
  []
  [T_cold]
    type = FVDirichletBC
    variable = T_fluid
    boundary = '3'
    value = 290
  []
[]

[AuxVariables]
  [U]
    order = CONSTANT
    family = MONOMIAL
    fv = true
  []
[]

[AuxKernels]
  [mag]
    type = VectorMagnitudeAux
    variable = U
    x = vel_x
    y = vel_y
    z = vel_z
  []
[]

[FluidProperties]
  [simple_fluid]
    type = SimpleFluidProperties
    bulk_modulus = 10000
    viscosity = 1
    density0 = 8.027
    thermal_expansion = 0
  []
[]

[FunctorMaterials]
  [functor_constants]
    type = ADGenericFunctorMaterial
    prop_names = 'cp k'
    prop_values = '${cp} ${k}'
  []
  [ins_fv]
    type = INSFVEnthalpyFunctorMaterial
    temperature = 'T_fluid'
    rho = ${rho}
  []
[]

[Functions]
  [velocity_z]
    type = ParsedFunction
    expression = (8.9049*10^5)*sin(t*((2*pi)/(12*10^-9)))
  []
[]

[Postprocessors]
  [element_T]
    type = ElementAverageValue
    variable = T_fluid
    execute_on = 'LINEAR'
  []
  [nodal_z]
    type = ElementAverageValue
    variable = vel_z
    execute_on = 'LINEAR'
  []
  [element_pressure]
    type = ElementAverageValue
    variable = pressure
    execute_on = 'LINEAR'
  []
[]

#[Preconditioning]
#  [SMP]
#    type = SMP
#    full = true
#  []
#[]

[Executioner]
  type = Transient
  end_time = 1.2e-8
  dt = 5.0e-11
  l_max_its  =  1000 #controls iterations
  nl_rel_tol = 0.01
#  line_search = 'none'
  petsc_options_iname = '-snes_linesearch_damping'
  petsc_options_value = '0.75'
[]

[Outputs]
  exodus = true
  print_linear_residuals = true
  [csv]
    type = CSV
    execute_on = 'LINEAR'
  []
[]

[Debug]
  show_var_residual_norms = true
[]

'''
Output residual example is:

'''

    148 Linear |R| = 1.756055e+07
    149 Linear |R| = 1.745537e+07
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 150
    |residual|_2 of individual variables:
                   vel_x:    5942.44
                   vel_y:    5990.86
                   vel_z:    162517
                   pressure: 216.941
                   T_fluid:  2.13954e+07
14 Nonlinear |R| = 2.139600e+07
      0 Linear |R| = 2.139600e+07
      1 Linear |R| = 2.139600e+07
      2 Linear |R| = 2.139600e+07
      3 Linear |R| = 2.139600e+07

     28 Linear |R| = 2.139597e+07
     29 Linear |R| = 2.139596e+07
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 30
    |residual|_2 of individual variables:
                   vel_x:    5942.44
                   vel_y:    5990.86
                   vel_z:    162517
                   pressure: 216.941
                   T_fluid:  2.13954e+07
15 Nonlinear |R| = 2.139600e+07
      0 Linear |R| = 2.139600e+07
      1 Linear |R| = 2.139600e+07
      2 Linear |R| = 2.139600e+07

'''

Thanks

### Additional information

Mesh size and type:
 Reynolds number:
 Discretization (finite element CG/DG, finite volume, etc):
 Models (turbulence, porous media, etc):
 Solver method (fully coupled, segregated, multiapps, etc):
 Base input you started from:

**Comment by GiudGiud** [comment]
Hello

An incompressible formulation won't do much for a shockwave, it will just propagate instantly and change the mean pressure of the system. 

For preconditioning, you can use LU with a nonzero factor shift for the Newton solve method. 

With average advected quantity interpolation you can expect checkerboarding of the pressure, we would recommend using upwind here

**Reply by Spudman105** [reply]
Would a weakly compressible be better then to allow for some response to the load?

**Reply by Spudman105** [reply]
Additionally I made the changes you recommended. I got these errors back and looked in documentation but couldn't find anything about how t solve this error code. 
'''
[0]PETSC ERROR: Residual norm computed by GMRES recursion formula 36319. is far from the computed residual norm 1.37661e+12 at restart, residual norm at start of cycle 7.29559e+08

Residual norm computed by GMRES recursion formula 36319. is far from the computed residual norm 1.37661e+12 at restart, residual norm at start of cycle 7.29559e+08
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 3
libMesh terminating:
'''

Any help is appreciated,
Thanks

**Reply by GiudGiud** [reply]
Weakly compressible has the same issues as incompressible.
We use the compressible formulation for any fast flow or shock problem. It is less mature though

**Reply by GiudGiud** [reply]
I don't know about the errors. You ll have to look at petsc or libmesh to see why that may be

**Reply by GiudGiud** [reply]
@lindsayad have you hit that one before?
Looking there https://petsc.org/release/src/ksp/ksp/impls/gmres/gmres.c.html it seems this error is a sign of non-convergence of GMRES

**Reply by lindsayad** [reply]
I believe I have but it's been a long time. You should run with `Executioner/solve_type = NEWTON` and then you should only be getting a single KSP iteration if you are using LU. Also use `-pc_factor_mat_solver_type mumps` to avoid using the PETSc native implementation (which is not very accureate) in serial
